File: scala/meetup/src/main/scala/Meetup.scala
import java.util.GregorianCalendar
import java.time.{LocalDate, DayOfWeek}
import java.time.temporal.TemporalAdjusters

object Meetup {
  import DayOfWeek._
  val Mon = MONDAY
  val Tue = TUESDAY
  val Wed = WEDNESDAY
  val Thu = THURSDAY
  val Fri = FRIDAY
  val Sat = SATURDAY
  val Sun = SUNDAY

  implicit class LocalDateHelpers(localDate: LocalDate) {
    def toCalendar =
      new GregorianCalendar(localDate.getYear, localDate.getMonthValue-1, localDate.getDayOfMonth)
    def isTeenth = 13 <= localDate.getDayOfMonth && localDate.getDayOfMonth < 20
  }

  def apply(m:Int, y:Int) = new Meetup(m, y)
}

class Meetup(m: Int, y: Int) {
  import Meetup._

  def dates(day: DayOfWeek) = {
    val first = LocalDate.of(y, m, 1)
    val firstOfNext = first.plusMonths(1)

    (1 until 6)
      .map(TemporalAdjusters.dayOfWeekInMonth(_:Int, day))
      .map(first.`with`)
      .filter(_.compareTo(firstOfNext) < 0)
      .toIndexedSeq
  }

  def teenth(day: DayOfWeek) = dates(day).find(_.isTeenth).get.toCalendar

  def first(day: DayOfWeek) = dates(day)(0).toCalendar

  def second(day: DayOfWeek) = dates(day)(1).toCalendar

  def third(day: DayOfWeek) = dates(day)(2).toCalendar

  def fourth(day: DayOfWeek) = dates(day)(3).toCalendar

  def last(day: DayOfWeek) = dates(day).last.toCalendar
}

File: java/meetup/src/main/java/Meetup.java
import org.joda.time.DateTime;

import java.util.function.Predicate;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Meetup {
    private final DateTime firstOfMonth;
    private final DateTime firstOfNextMonth;

    public Meetup(int month, int year) {
        this.firstOfMonth = new DateTime(year, month, 1, 0, 0);
        this.firstOfNextMonth = firstOfMonth.plusMonths(1);
    }

    private final Predicate<DateTime> isTeenth = (d)-> 13 <= d.getDayOfMonth() && d.getDayOfMonth() < 20;

    public final Stream<DateTime> dates(int dayOfWeek) {
        return IntStream.rangeClosed(0, 5)
                .mapToObj(week-> firstOfMonth.plusWeeks(week))
                .map(dateTime -> dateTime.withDayOfWeek(dayOfWeek))
                .filter(dateTime -> dateTime.compareTo(firstOfMonth) >= 0)
                .filter(dateTime-> dateTime.compareTo(firstOfNextMonth) < 0);
    }

    public DateTime day(int dayOfWeek, MeetupSchedule meetupSchedule) {
        switch (meetupSchedule) {
            case TEENTH:
                return dates(dayOfWeek).filter(isTeenth).findFirst().get();
            case FIRST:
                return dates(dayOfWeek).findFirst().get();
            case SECOND:
                return dates(dayOfWeek).skip(1).findFirst().get();
            case THIRD:
                return dates(dayOfWeek).skip(2).findFirst().get();
            case FOURTH:
                return dates(dayOfWeek).skip(3).findFirst().get();
            case LAST:
                return dates(dayOfWeek).reduce((left, right)-> right).get();
            default:
                return null;
        }
    }
}
